### PR TITLE
ecl: 20.4.24 -> 21.2.1, enable Darwin

### PR DESCRIPTION
**Pull request by purcell**
###### Motivation for this change

- Incorporates the latest upstream release
- Removes now-defunct patches
- Enable builds on non-Linux Unixes, ie. Darwin

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
      - No, could not be built locally previously. Looks reasonable though, given that it includes gcc.
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by purcell** [comment]
Seems to break `maxima-ecl`, at least on Darwin: `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` says
```
error: --- Error ------------------------------------------------------------------------------------------------- nix
while setting up the build environment: getting attributes of path '/usr/lib/libSystem.B.dylib': No such file or directory
1 package failed to build:
maxima-ecl
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112718` at 8628c902 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/b822bf24cc652b510ec161f57b1c58e5">ecl</a></li></li>
    <li><a href="https://gist.github.com/12c1199bfa10fb5068b50a107ba75120">eql</a></li></li>
    <li><a href="https://gist.github.com/3cd2e7d95118f5a6b5be0ac5c8a8b073">maxima-ecl</a></li></li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112718` at 8628c902 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>sage</li>
    <li>sageWithDoc</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/dd3b012525c0d218b2de5ead085613a2">ecl</a></li></li>
    <li><a href="https://gist.github.com/423a87dd280b27a41fedf5eb99196f1f">eql</a></li></li>
    <li><a href="https://gist.github.com/919d1479c2d1027c39c76ddc645405a4">maxima-ecl</a></li></li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112718` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>ecl</li>
<li>maxima-ecl</li>
</ul>
</details>

File: pkgs/development/compilers/ecl/default.nix [hunk 1]
```diff
@@ -36,38 +36,12 @@ stdenv.mkDerivation {
...
   patches = [
-    # https://trac.sagemath.org/ticket/22191#comment:237
-    (fetchurl {
-      name = "ECL_WITH_LISP_FPE.patch";
-      url = "https://git.sagemath.org/sage.git/plain/build/pkgs/ecl/patches/ECL_WITH_LISP_FPE.patch?h=9.2";
-      sha256 = "0b194613avcmzr1k9gq725z41wdkg5rsa0q21kdw050iqpprcj1c";
-    })
-
     # https://gitlab.com/embeddable-common-lisp/ecl/-/merge_requests/1
     (fetchurl {
```
[the remark this answers is not in the document]
Comment (purcell) [reply pkgs/development/compilers/ecl/default.nix:40]: Sure, I can potentially clean up that pre-existing stylistic mistake along the way.

[the remark this answers is not in the document]
Comment (purcell) [reply pkgs/development/compilers/ecl/default.nix]: Done

File: pkgs/development/compilers/ecl/default.nix [hunk 2]
```diff
@@ -36,38 +36,12 @@ stdenv.mkDerivation {
...
-      url = "https://git.sagemath.org/sage.git/plain/build/pkgs/ecl/patches/ECL_WITH_LISP_FPE.patch?h=9.2";
-      sha256 = "0b194613avcmzr1k9gq725z41wdkg5rsa0q21kdw050iqpprcj1c";
-    })
-
     # https://gitlab.com/embeddable-common-lisp/ecl/-/merge_requests/1
-    (fetchurl {
+    (fetchpatch {
       name = "write_error.patch";
       url = "https://git.sagemath.org/sage.git/plain/build/pkgs/ecl/patches/write_error.patch?h=9.2";
       sha256 = "1lvdvh77blnxp0zbd27dsbq1ljkb5qblabf1illszn4j7qgq88fh";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/ecl/default.nix:43]: I think you need to check if the hash changed. It most likely did.

Comment (purcell) [reply pkgs/development/compilers/ecl/default.nix:43]: Good call, fixed now, thanks.

**Pull request merged**